File: internal/auth/sigv4.go
// Copyright 2026 vs3-neo Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package auth implements AWS Signature Version 4 request signing used by
// S3, supporting both header-based (Authorization) and query-based
// (presigned URL) authentication.
package auth

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

const (
	algorithm = "AWS4-HMAC-SHA256"
	service   = "s3"
	// unsignedPayload is used when a request body hash is unavailable or the
	// client used the S3 "unsigned payload" mode (typical for presigned URLs).
	unsignedPayload = "UNSIGNED-PAYLOAD"
)

// ErrNoAuth is returned when the request carries no signature.
var ErrNoAuth = errors.New("no authentication provided")

// Credential identifies an authenticated caller.
type Credential struct {
	AccessKey string
}

// Provider resolves the secret key for an access key.
type Provider interface {
	SecretKey(accessKey string) (string, bool)
}

// Verifier validates SigV4-signed requests against a Provider.
type Verifier struct {
	Provider Provider
	// Clock is used for presigned URL expiration checks (tests override it).
	Clock func() time.Time
	// AllowAnonymous permits unsigned requests.
	AllowAnonymous bool
}

// NewVerifier returns a Verifier for p.
func NewVerifier(p Provider) *Verifier {
	return &Verifier{Provider: p, Clock: time.Now}
}

// Verify checks the request's SigV4 signature. It returns the caller
// credential and whether the request was signed or anonymous.
func (v *Verifier) Verify(r *http.Request) (Credential, bool, error) {
	authz := r.Header.Get("Authorization")
	if strings.HasPrefix(authz, algorithm+" ") {
		cred, err := v.verifyHeader(r, authz)
		return cred, true, err
	}
	// Query-based auth (presigned URL).
	if r.URL.Query().Get("X-Amz-Algorithm") == algorithm {
		cred, err := v.verifyQuery(r)
		return cred, true, err
	}
	if v.AllowAnonymous {
		return Credential{}, false, nil
	}
	return Credential{}, false, ErrNoAuth
}

// ---- header-based auth ----

type authParts struct {
	accessKey     string
	date          string
	region        string
	service       string
	signedHeaders []string
	signature     string
}

func parseAuthHeader(authz string) (*authParts, error) {
	// "AWS4-HMAC-SHA256 Credential=AK/date/region/s3/aws4_request, SignedHeaders=..., Signature=..."
	body := strings.TrimPrefix(authz, algorithm+" ")
	parts := map[string]string{}
	for _, kv := range strings.Split(body, ",") {
		kv = strings.TrimSpace(kv)
		idx := strings.Index(kv, "=")
		if idx < 0 {
			continue
		}
		parts[strings.TrimSpace(kv[:idx])] = strings.TrimSpace(kv[idx+1:])
	}
	cred := parts["Credential"]
	credParts := strings.Split(cred, "/")
	if len(credParts) != 5 || credParts[4] != "aws4_request" {
		return nil, errors.New("auth: malformed credential scope")
	}
	ap := &authParts{
		accessKey: credParts[0],
		date:      credParts[1],
		region:    credParts[2],
		service:   credParts[3],
		signature: parts["Signature"],
	}
	if sh := parts["SignedHeaders"]; sh != "" {
		ap.signedHeaders = strings.Split(sh, ";")
	}
	return ap, nil
}

func (v *Verifier) verifyHeader(r *http.Request, authz string) (Credential, error) {
	ap, err := parseAuthHeader(authz)
	if err != nil {
		return Credential{}, err
	}
	if ap.service != service {
		return Credential{}, fmt.Errorf("auth: unsupported service %q", ap.service)
	}
	secret, ok := v.Provider.SecretKey(ap.accessKey)
	if !ok {
		return Credential{}, errors.New("auth: unknown access key")
	}
	amzDate := r.Header.Get("x-amz-date")
	if amzDate == "" {
		amzDate = ap.date + "T000000Z"
	}
	canonReq := buildCanonicalRequest(r, ap.signedHeaders, headerPayloadHash(r))
	stringToSign := buildStringToSign(amzDate, ap.date, ap.region, canonReq)
	signingKey := signingKey(secret, ap.date, ap.region)
	expected := hmacHex(signingKey, stringToSign)
	if !hmac.Equal([]byte(expected), []byte(ap.signature)) {
		return Credential{}, errors.New("auth: signature mismatch")
	}
	return Credential{AccessKey: ap.accessKey}, nil
}

func headerPayloadHash(r *http.Request) string {
	if h := r.Header.Get("x-amz-content-sha256"); h != "" {
		return h
	}
	return unsignedPayload
}

// ---- query-based (presigned) auth ----

func (v *Verifier) verifyQuery(r *http.Request) (Credential, error) {
	q := r.URL.Query()
	cred := q.Get("X-Amz-Credential")
	credParts := strings.Split(cred, "/")
	if len(credParts) != 5 || credParts[4] != "aws4_request" {
		return Credential{}, errors.New("auth: malformed presigned credential")
	}
	accessKey, date, region, svc := credParts[0], credParts[1], credParts[2], credParts[3]
	if svc != service {
		return Credential{}, fmt.Errorf("auth: unsupported service %q", svc)
	}
	expiresStr := q.Get("X-Amz-Expires")
	var expires time.Duration
	if expiresStr != "" {
		var secs int64
		if _, err := fmt.Sscanf(expiresStr, "%d", &secs); err != nil {
			return Credential{}, errors.New("auth: bad X-Amz-Expires")
		}
		expires = time.Duration(secs) * time.Second
	}
	amzDate, err := time.Parse("20060102T150405Z", q.Get("X-Amz-Date"))
	if err != nil {
		return Credential{}, errors.New("auth: bad X-Amz-Date")
	}
	now := time.Now()
	if v.Clock != nil {
		now = v.Clock()
	}
	if now.After(amzDate.Add(expires)) {
		return Credential{}, errors.New("auth: presigned URL expired")
	}
	if now.Before(amzDate.Add(-15 * time.Minute)) {
		return Credential{}, errors.New("auth: presigned URL too far in the future")
	}
	secret, ok := v.Provider.SecretKey(accessKey)
	if !ok {
		return Credential{}, errors.New("auth: unknown access key")
	}
	signedHeaders := []string{}
	if sh := q.Get("X-Amz-SignedHeaders"); sh != "" {
		signedHeaders = strings.Split(sh, ";")
	}
	canonReq := buildCanonicalRequest(r, signedHeaders, unsignedPayload)
	stringToSign := buildStringToSign(q.Get("X-Amz-Date"), date, region, canonReq)
	key := signingKey(secret, date, region)
	expected := hmacHex(key, stringToSign)
	if !hmac.Equal([]byte(expected), []byte(q.Get("X-Amz-Signature"))) {
		return Credential{}, errors.New("auth: signature mismatch")
	}
	return Credential{AccessKey: accessKey}, nil
}

// ---- canonical request construction ----

// buildCanonicalRequest assembles the AWS SigV4 canonical request. For query
// auth the X-Amz-Signature parameter is excluded, as required by AWS.
func buildCanonicalRequest(r *http.Request, signedHeaders []string, payloadHash string) string {
	method := r.Method
	canonicalURI := canonicalURI(r.URL)
	canonicalQuery := canonicalQueryString(r.URL, isQueryAuth(r))
	canonicalHeaders, headersList := canonicalHeadersString(r, signedHeaders)
	canonical := strings.Join([]string{
		method,
		canonicalURI,
		canonicalQuery,
		canonicalHeaders,
		headersList,
		payloadHash,
	}, "\n")
	return canonical
}

func isQueryAuth(r *http.Request) bool {
	return r.URL.Query().Get("X-Amz-Algorithm") == algorithm
}

// canonicalURI returns the URI-encoded path as sent by the client (S3 does
// not double-encode, so the client's own encoding is the canonical form).
func canonicalURI(u *url.URL) string {
	p := u.EscapedPath()
	if p == "" {
		return "/"
	}
	return p
}

// canonicalQueryString sorts raw query pairs (excluding X-Amz-Signature for
// query auth) and joins them as name=value.
func canonicalQueryString(u *url.URL, excludeSignature bool) string {
	raw := u.RawQuery
	if raw == "" {
		return ""
	}
	type kv struct{ k, v string }
	var pairs []kv
	for _, pair := range strings.Split(raw, "&") {
		if pair == "" {
			continue
		}
		idx := strings.Index(pair, "=")
		var k, v string
		if idx < 0 {
			k, v = pair, ""
		} else {
			k, v = pair[:idx], pair[idx+1:]
		}
		if excludeSignature && k == "X-Amz-Signature" {
			continue
		}
		pairs = append(pairs, kv{k, v})
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].k != pairs[j].k {
			return pairs[i].k < pairs[j].k
		}
		return pairs[i].v < pairs[j].v
	})
	var sb strings.Builder
	for i, p := range pairs {
		if i > 0 {
			sb.WriteByte('&')
		}
		sb.WriteString(p.k)
		sb.WriteByte('=')
		sb.WriteString(p.v)
	}
	return sb.String()
}

// canonicalHeadersString builds the sorted header block and the signed
// headers list (already sorted).
func canonicalHeadersString(r *http.Request, signedHeaders []string) (headersBlock, list string) {
	names := make([]string, len(signedHeaders))
	copy(names, signedHeaders)
	sort.Strings(names)

	// Lower-case + trim, preserving original header order from the request.
	values := map[string][]string{}
	for name := range r.Header {
		values[strings.ToLower(name)] = r.Header.Values(name)
	}
	// x-amz-date is often absent from header list but present in request.
	lowerHeaders := []string{}
	for _, n := range names {
		l := strings.ToLower(strings.TrimSpace(n))
		lowerHeaders = append(lowerHeaders, l)
	}
	var sb strings.Builder
	for _, name := range lowerHeaders {
		vs := values[name]
		if len(vs) == 0 {
			continue
		}
		val := strings.Join(vs, ",")
		// Collapse sequential spaces, per AWS spec.
		val = collapseSpaces(val)
		sb.WriteString(name)
		sb.WriteByte(':')
		sb.WriteString(val)
		sb.WriteByte('\n')
	}
	return sb.String(), strings.Join(lowerHeaders, ";")
}

func collapseSpaces(s string) string {
	trimmed := strings.TrimSpace(s)
	parts := strings.Fields(trimmed)
	return strings.Join(parts, " ")
}

func buildStringToSign(amzDate, date, region, canonicalRequest string) string {
	h := sha256.Sum256([]byte(canonicalRequest))
	scope := date + "/" + region + "/" + service + "/aws4_request"
	return strings.Join([]string{
		algorithm,
		amzDate,
		scope,
		hex.EncodeToString(h[:]),
	}, "\n")
}

func signingKey(secret, date, region string) []byte {
	kDate := hmacSHA256([]byte("AWS4"+secret), date)
	kRegion := hmacSHA256(kDate, region)
	kService := hmacSHA256(kRegion, service)
	return hmacSHA256(kService, "aws4_request")
}

func hmacSHA256(key []byte, data string) []byte {
	m := hmac.New(sha256.New, key)
	m.Write([]byte(data))
	return m.Sum(nil)
}

func hmacHex(key []byte, data string) string {
	return hex.EncodeToString(hmacSHA256(key, data))
}

// ---- Presigner (mint presigned URLs) ----

// Presigner creates SigV4 presigned URLs for any S3-style request.
type Presigner struct {
	AccessKey string
	SecretKey string
	Region    string
	// Clock for testing.
	Clock func() time.Time
}

// Presign returns a URL for method against `path` (e.g. "/bucket/key") with
// the given query parameters, valid for expires seconds.
func (p *Presigner) Presign(method, host, path string, query url.Values, expires int64) (string, error) {
	now := time.Now()
	if p.Clock != nil {
		now = p.Clock()
	}
	amzDate := now.UTC().Format("20060102T150405Z")
	date := amzDate[:8]
	scope := date + "/" + p.Region + "/" + service + "/aws4_request"

	if query == nil {
		query = url.Values{}
	}
	query.Set("X-Amz-Algorithm", algorithm)
	query.Set("X-Amz-Credential", p.AccessKey+"/"+scope)
	query.Set("X-Amz-Date", amzDate)
	query.Set("X-Amz-Expires", fmt.Sprintf("%d", expires))
	query.Set("X-Amz-SignedHeaders", "host")

	// Build canonical request excluding the signature itself.
	fakeURL := &url.URL{Path: path, RawQuery: query.Encode()}
	// canonical query: sort + encode with %20, exclude X-Amz-Signature.
	canonQuery := encodeSortedQuery(query)
	canonReq := strings.Join([]string{
		method,
		encodePath(path),
		canonQuery,
		"host:" + host + "\n",
		"host",
		unsignedPayload,
	}, "\n")
	stringToSign := buildStringToSign(amzDate, date, p.Region, canonReq)
	sig := hmacHex(signingKey(p.SecretKey, date, p.Region), stringToSign)
	query.Set("X-Amz-Signature", sig)

	u := url.URL{Scheme: "http", Host: host, Path: path, RawQuery: encodeSortedQuery(query)}
	return u.String(), nil
}

// encodeSortedQuery returns RFC3986-encoded query string sorted by key.
func encodeSortedQuery(q url.Values) string {
	keys := make([]string, 0, len(q))
	for k := range q {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	var sb strings.Builder
	for i, k := range keys {
		vals := q[k]
		sort.Strings(vals)
		for _, v := range vals {
			if sb.Len() > 0 {
				sb.WriteByte('&')
			}
			sb.WriteString(encodeComponent(k))
			sb.WriteByte('=')
			sb.WriteString(encodeComponent(v))
		}
		_ = i
	}
	return sb.String()
}

// encodeComponent percent-encodes per RFC3986 (space -> %20).
func encodeComponent(s string) string {
	return strings.ReplaceAll(url.QueryEscape(s), "+", "%20")
}

// encodePath percent-encodes each path segment while preserving "/".
func encodePath(p string) string {
	segments := strings.Split(p, "/")
	for i, seg := range segments {
		if seg != "" {
			segments[i] = encodeComponent(seg)
		}
	}
	return strings.Join(segments, "/")
}

File: src/auth/sigv4.js
// Copyright 2026 vs3-neo Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import crypto from 'node:crypto';

export const ALGORITHM = 'AWS4-HMAC-SHA256';
export const SERVICE = 's3';
const UNSIGNED_PAYLOAD = 'UNSIGNED-PAYLOAD';

// S3-compatible error codes for auth failures.
export class AuthError extends Error {
  constructor(code, message, status = 403) {
    super(message);
    this.code = code;
    this.status = status;
  }
}

const sigMismatch = () =>
  new AuthError('SignatureDoesNotMatch', 'The request signature we calculated does not match the signature you provided.');
const noAuth = () => new AuthError('AccessDenied', 'Access Denied. No credentials provided.', 403);
const expired = () => new AuthError('AccessDenied', 'Request has expired', 403);

function hmac(key, data) {
  return crypto.createHmac('sha256', key).update(data).digest();
}

function sha256hex(data) {
  return crypto.createHash('sha256').update(data).digest('hex');
}

// Parse AWS SigV4 basic date "YYYYMMDDTHHMMSSZ" into a Date. This is not the
// ISO8601 form that `new Date()` accepts, so it must be parsed manually.
export function parseAmzDate(s) {
  const m = /^(\d{4})(\d{2})(\d{2})T(\d{2})(\d{2})(\d{2})Z$/.exec(s || '');
  if (!m) return new Date(NaN);
  return new Date(Date.UTC(+m[1], +m[2] - 1, +m[3], +m[4], +m[5], +m[6]));
}

function hex(buf) {
  return buf.toString('hex');
}

function signingKey(secret, date, region) {
  const kDate = hmac('AWS4' + secret, date);
  const kRegion = hmac(kDate, region);
  const kService = hmac(kRegion, SERVICE);
  return hmac(kService, 'aws4_request');
}

// ---- canonical request helpers ----

// RFC3986 percent-encode (space -> %20).
function enc(s) {
  return encodeURIComponent(s).replace(/[!'()*]/g, (c) => '%' + c.charCodeAt(0).toString(16).toUpperCase());
}

function canonicalURI(rawPath) {
  if (!rawPath || rawPath === '') return '/';
  return rawPath
    .split('/')
    .map((seg) => (seg === '' ? '' : enc(seg)))
    .join('/');
}

// canonicalQueryString sorts raw name=value pairs. For presigned (query) auth
// the X-Amz-Signature parameter is excluded.
function canonicalQueryString(queryParts, excludeSignature) {
  let pairs = queryParts.filter((p) => p.name !== '');
  if (excludeSignature) {
    pairs = pairs.filter((p) => p.name !== 'X-Amz-Signature');
  }
  pairs.sort((a, b) => (a.name + '\u0000' + a.value).localeCompare(b.name + '\u0000' + b.value));
  return pairs.map((p) => `${p.name}=${p.value}`).join('&');
}

// Parse raw query string into {name, value} pairs preserving original encoding.
function parseQuery(rawQuery) {
  if (!rawQuery) return [];
  const pairs = [];
  for (const pair of rawQuery.split('&')) {
    if (pair === '') continue;
    const idx = pair.indexOf('=');
    if (idx < 0) pairs.push({ name: pair, value: '' });
    else pairs.push({ name: pair.slice(0, idx), value: pair.slice(idx + 1) });
  }
  return pairs;
}

// Build canonical headers block. headers is the lowercased map from Node.
function canonicalHeaders(headers, signedHeaders) {
  const sorted = signedHeaders.slice().sort();
  const lines = [];
  const list = [];
  for (const name of sorted) {
    const value = headers[name];
    if (value === undefined) continue;
    // Multi-value headers are comma-joined; collapse whitespace per AWS.
    const joined = (Array.isArray(value) ? value : [value])
      .map((v) => String(v).trim().replace(/\s+/g, ' '))
      .join(',');
    lines.push(`${name}:${joined}\n`);
    list.push(name);
  }
  return { block: lines.join(''), list: list.join(';') };
}

function buildStringToSign(amzDate, scope, canonicalRequest) {
  return [
    ALGORITHM,
    amzDate,
    scope,
    sha256hex(canonicalRequest),
  ].join('\n');
}

function buildCanonicalRequest(req, method, queryParts, signedHeaders, payloadHash) {
  const { block, list } = canonicalHeaders(req.headers, signedHeaders);
  const canonical = [
    method,
    canonicalURI(req.path),
    canonicalQueryString(queryParts, req.isQueryAuth),
    block,
    list,
    payloadHash,
  ].join('\n');
  return canonical;
}

// ---- request verification ----

function headerPayloadHash(req) {
  const v = req.headers['x-amz-content-sha256'];
  return v || UNSIGNED_PAYLOAD;
}

function parseAuthHeader(authz) {
  const body = authz.slice(ALGORITHM.length + 1);
  const parts = {};
  for (const kv of body.split(',')) {
    const idx = kv.indexOf('=');
    if (idx < 0) continue;
    parts[kv.slice(0, idx).trim()] = kv.slice(idx + 1).trim();
  }
  const credParts = (parts.Credential || '').split('/');
  if (credParts.length !== 5 || credParts[4] !== 'aws4_request') {
    throw new AuthError('AuthorizationHeaderMalformed', 'The authorization header is malformed.', 400);
  }
  return {
    accessKey: credParts[0],
    date: credParts[1],
    region: credParts[2],
    service: credParts[3],
    signedHeaders: parts.SignedHeaders ? parts.SignedHeaders.split(';') : [],
    signature: parts.Signature || '',
  };
}

function verifyHeaderAuth(req, provider) {
  const ap = parseAuthHeader(req.headers.authorization);
  if (ap.service !== SERVICE) {
    throw new AuthError('AuthorizationHeaderMalformed', `Unsupported service ${ap.service}`, 400);
  }
  const secret = provider.secretKey(ap.accessKey);
  if (!secret) throw new AuthError('InvalidAccessKeyId', 'The AWS Access Key Id you provided does not exist in our records.');
  const amzDate = req.headers['x-amz-date'] || `${ap.date}T000000Z`;
  const canonical = buildCanonicalRequest(req, req.method, req.queryParts, ap.signedHeaders, headerPayloadHash(req));
  const scope = `${ap.date}/${ap.region}/${SERVICE}/aws4_request`;
  const sts = buildStringToSign(amzDate, scope, canonical);
  const expected = hex(hmac(signingKey(secret, ap.date, ap.region), sts));
  if (!timingSafeEqualHex(expected, ap.signature)) throw sigMismatch();
  return ap.accessKey;
}

function verifyQueryAuth(req, provider, clock) {
  const q = req.queryParams; // decoded params
  const credParts = (q['X-Amz-Credential'] || '').split('/');
  if (credParts.length !== 5 || credParts[4] !== 'aws4_request') {
    throw new AuthError('AuthorizationQueryParametersError', 'Query-string authentication version 4 requires the X-Amz-Algorithm, X-Amz-Credential, X-Amz-Signature, X-Amz-Date, X-Amz-SignedHeaders, and X-Amz-Expires parameters.', 400);
  }
  const [accessKey, date, region, service] = credParts;
  if (service !== SERVICE) {
    throw new AuthError('AuthorizationQueryParametersError', `Unsupported service ${service}`, 400);
  }
  const expires = parseInt(q['X-Amz-Expires'] || '0', 10);
  const amzDate = parseAmzDate(q['X-Amz-Date']);
  const now = clock ? clock() : new Date();
  if (Number.isNaN(amzDate.getTime())) {
    throw new AuthError('AccessDenied', 'Invalid date in query string', 403);
  }
  if (now.getTime() > amzDate.getTime() + expires * 1000) throw expired();
  const secret = provider.secretKey(accessKey);
  if (!secret) throw new AuthError('InvalidAccessKeyId', 'The AWS Access Key Id you provided does not exist in our records.');
  const signedHeaders = (q['X-Amz-SignedHeaders'] || '').split(';').filter(Boolean);
  const canonical = buildCanonicalRequest(req, req.method, req.queryParts, signedHeaders, UNSIGNED_PAYLOAD);
  const scope = `${date}/${region}/${SERVICE}/aws4_request`;
  const sts = buildStringToSign(q['X-Amz-Date'], scope, canonical);
  const expected = hex(hmac(signingKey(secret, date, region), sts));
  if (!timingSafeEqualHex(expected, q['X-Amz-Signature'] || '')) throw sigMismatch();
  return accessKey;
}

function timingSafeEqualHex(a, b) {
  const ab = Buffer.from(a, 'hex');
  const bb = Buffer.from(b, 'hex');
  if (ab.length !== bb.length) return false;
  return crypto.timingSafeEqual(ab, bb);
}

// ---- public verifier ----

// req is a normalized request object:
//   { method, path, rawQuery, queryParts, queryParams, headers, isQueryAuth }
// provider: { secretKey(accessKey) -> secret|undefined }
// opts: { anonymous, clock }
// Returns { accessKey, signed }.
export function verifyRequest(req, provider, opts = {}) {
  const authz = req.headers.authorization || '';
  if (authz.startsWith(ALGORITHM + ' ')) {
    const accessKey = verifyHeaderAuth(req, provider);
    return { accessKey, signed: true };
  }
  if (req.isQueryAuth) {
    const accessKey = verifyQueryAuth(req, provider, opts.clock);
    return { accessKey, signed: true };
  }
  if (opts.anonymous) return { accessKey: null, signed: false };
  throw noAuth();
}

// ---- Presigner (mint presigned URLs) ----

export class Presigner {
  constructor({ accessKey, secretKey, region = 'us-east-1', clock } = {}) {
    this.accessKey = accessKey;
    this.secretKey = secretKey;
    this.region = region;
    this.clock = clock;
  }

  // Presign returns a URL for method against `path` (e.g. "/bucket/key").
  // query is a plain object of extra params (encoded properly).
  presign(method, host, path, query = {}, expires = 3600) {
    const now = this.clock ? this.clock() : new Date();
    const amzDate = now.toISOString().replace(/[-:]/g, '').replace(/\.\d{3}Z$/, 'Z');
    const date = amzDate.slice(0, 8);
    const scope = `${date}/${this.region}/${SERVICE}/aws4_request`;

    const params = { ...query };
    params['X-Amz-Algorithm'] = ALGORITHM;
    params['X-Amz-Credential'] = `${this.accessKey}/${scope}`;
    params['X-Amz-Date'] = amzDate;
    params['X-Amz-Expires'] = String(expires);
    params['X-Amz-SignedHeaders'] = 'host';

    const sortedKeys = Object.keys(params).sort();
    const canonicalQuery = sortedKeys
      .map((k) => `${enc(k)}=${enc(params[k])}`)
      .join('&');

    const canonical = [
      method,
      canonicalURI(path),
      canonicalQuery,
      `host:${host}\n`,
      'host',
      UNSIGNED_PAYLOAD,
    ].join('\n');
    const sts = buildStringToSign(amzDate, scope, canonical);
    const signature = hex(hmac(signingKey(this.secretKey, date, this.region), sts));

    const q = sortedKeys.map((k) => `${enc(k)}=${enc(params[k])}`).join('&');
    const qs = `${q}&X-Amz-Signature=${signature}`;
    return `http://${host}${path}?${qs}`;
  }
}

// ---- client-side signer (used by the Go/Node test clients & SDK parity) ----

// signHeaders returns the Authorization header for a normal (header-based)
// signed request.
export function signRequest({ method, host, path, query = {}, headers = {}, bodyHash, secretKey, accessKey, region = 'us-east-1', amzDate }) {
  const now = new Date();
  const ts = (amzDate || now.toISOString().replace(/[-:]/g, '').replace(/\.\d{3}Z$/, 'Z'));
  const date = ts.slice(0, 8);
  const scope = `${date}/${region}/${SERVICE}/aws4_request`;
  const payloadHash = bodyHash || UNSIGNED_PAYLOAD;

  const signedHeaderNames = Object.keys(headers)
    .map((k) => k.toLowerCase())
    .sort();
  signedHeaderNames.push('host');
  const sortedUnique = [...new Set(signedHeaderNames)].sort();

  // Canonical headers from the provided set (plus host). Header names are
  // normalized to lowercase so lookups against the sorted signed list work
  // even when callers pass mixed-case names (e.g. "Content-Type").
  const allHeaders = { host };
  for (const [k, v] of Object.entries(headers)) {
    allHeaders[k.toLowerCase()] = v;
  }
  const lines = [];
  for (const name of sortedUnique) {
    const value = allHeaders[name];
    if (value === undefined) continue;
    lines.push(`${name}:${String(value).trim().replace(/\s+/g, ' ')}\n`);
  }
  const canonicalHeadersBlock = lines.join('');
  const signedHeadersList = sortedUnique.join(';');

  const queryPairs = Object.entries(query)
    .map(([k, v]) => ({ name: enc(k), value: enc(v) }))
    .sort((a, b) => (a.name + '\u0000' + a.value).localeCompare(b.name + '\u0000' + b.value));
  const canonicalQuery = queryPairs.map((p) => `${p.name}=${p.value}`).join('&');

  const canonical = [
    method,
    canonicalURI(path),
    canonicalQuery,
    canonicalHeadersBlock,
    signedHeadersList,
    payloadHash,
  ].join('\n');
  const sts = buildStringToSign(ts, scope, canonical);
  const signature = hex(hmac(signingKey(secretKey, date, region), sts));
  const authz = `${ALGORITHM} Credential=${accessKey}/${scope}, SignedHeaders=${signedHeadersList}, Signature=${signature}`;
  return { authorization: authz, 'x-amz-date': ts, 'x-amz-content-sha256': payloadHash };
}
